### ISSUE TITLE
Circular links cause stack overflow

### ISSUE THREAD

**Issue by nijotz** [contentful-labs/contentful_middleman#42]
/Users/nicktzaperas/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/contentful_middleman-a512465f271c/lib/contentful_middleman/mappers/base.rb:50:in `map_entry': stack level too deep (SystemStackError)

**Comment by nijotz** [comment]
This seems to fix it:
https://github.com/contentful-labs/contentful_middleman/pull/43
Pretty basic, but works

**Comment by nijotz** [comment]
Just came across something.  The existing code traverses a tree (really a graph, but in most cases a tree) pre-order.  By traversing pre-order until it comes across an entity it has already encountered, it can create lop-sided data structures.  I have an entity that links to four other entities and only two of them have the full data, the other two are just IDs.  Traversing breadth-first may be more ideal?

Anyway, looks like the this issue came up because my `:include => 1` in the cda_query doesn't seem to be working affecting the depth of returned entries.  contentful.rb is getting the `:include => `1, still need to look into it more.  So, contentful.rb is pulling what looks like 4 levels of linked entries, handing it to contentful_middleman and the mapper can easily find circular references what that many levels of linked entries.

If I find an issue with the :include possibly being ignored, I'll report it.

**Comment by nijotz** [comment]
Ahhh, okay.  So, the `:include => 1` is working, but the ResourceBuilder class in contentful.rb uses `replace_children`, `replace_children_array`, and `replace_links_with_known_resources` which creates circular links.

I'm going to redo pull request #43 to do a breadth-first search, stopping when it gets to a resource that's already been mapped.  I'm not proposing that this is the ideal solution for your projects.  But, it will work for my use-case and you can use it as a reference to understand the problem I ran into and a possible solution.

**Comment by nijotz** [comment]
I opened #44, which does a breadth-first search of entries.  This fixed the lop-sided data structures I mentioned [here](https://github.com/contentful-labs/contentful_middleman/issues/42#issuecomment-109420411)

**Comment by katperez** [comment]
Hi there. Not sure if it is just me but I am now getting something similar except it is with map_asset - 

/Users/katperez/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/contentful_middleman-1.1.0/lib/contentful_middleman/mappers/base.rb:42:in `map_asset': stack level too deep (SystemStackError)

**Comment by nijotz** [comment]
@katperez Ya, that looks like the same error I was getting.  I'm curious if the change I made fixes your problem as well.  Have you tried the fix in that PR?  I've been using it by changing my Gemfile to have this:

`gem "contentful_middleman", :git => "https://github.com/Instrument/contentful_middleman.git"`

**Comment by katperez** [comment]
@nijotz Well I added the asset id after line 43 and it seems to work now - 
```
def map_asset(asset)
        context       = Context.new
        context.id    = asset.id
        context.title = asset.title
        context.url   = asset.file.url

        context
      end

**Comment by nijotz** [comment]
Huh, interesting that *that* change would make the circular calls stop

**Comment by katperez** [comment]
@nijotz I have no clue

**Comment by pxlpnk** [comment]
We will look into this as soon as possible. Sorry for the delay

**Comment by neonichu** [comment]
This was fixed by #44, I believe.
